### PR TITLE
Cache blocking

### PR DESCRIPTION
**Pull request by navjotk**
Inviting review/comments on this.
This branch builds on https://github.com/opesci/devito/pull/17 so it should only be accepted after that.

### CODE REVIEW

**Comment by mlange05** [comment]
Now that PR #17 has landed can we please rebase this to make the diff readable?

**Comment by navjotk** [comment]
Rebased on master to make the diff readable. Please review.

File: devito/propagator.py [hunk 1]
```diff
@@ -11,6 +11,8 @@ def __init__(self, name, nt, shape, spc_border=0, forward=True, time_order=0):
         self.t = symbols("t")
         # We assume the dimensions are passed to us in the following order
         # var_order
+        self.cache_blocking = True
+        self.block_sizes = [10]*num_spac_dim
```
Comment (mlange05) [inline devito/propagator.py]: Where does the default block size come from? How would a user change the block size?

**Comment by mlange05** [comment]
Changes look good, but need testing. The branch also needs to be reconciled with master.

File: devito/propagator.py (hunk 1 above)
Comment (navjotk) [inline devito/propagator.py]: I expect this to come from the config file.

**Comment by navjotk** [comment]
Can you suggest a way to test this? I don't see how this could be tested apart from reading in the full C code as a string and testing equivalence with a hand-written string.

**Comment by ggorman** [comment]
In this I would suggest a regression test.
1. Run the non-blocked version and calculate the l2 norm of the solution
field
2. Run with blocking and again calculate the l2 norm of the solution field.
3. Assert the two l2 norms are the same (I think they should be exact
although different code might have minor changes to ordering of
operations).

Independently from this PR @mlouboutin is working on an MMS test which will
close the testing loop.

**Comment by navjotk** [comment]
@pp1336 Please keep your commits on a different branch. Let's keep the autotuning feature separate from cache blocking so this PR doesn't depend on your work.

**Comment by navjotk** [comment]
@sheino you reverted my last commit instead of your own

**Comment by sheino** [comment]
Should be alright now.

**Comment by mlange05** [comment]
Can we please just drop the last 4 commits to avoid an overly noisy commit history?

**Comment by navjotk** [comment]
Done the reset. Please review @mlange05

File: devito/propagator.py [hunk 2]
```diff
@@ -3,14 +3,20 @@
...
 
 
 class Propagator(object):
-    _ENV_VAR_PROFILE = "DEVITO_PROFILE"
-
-    def __init__(self, name, nt, shape, spc_border=0, forward=True, time_order=0, openmp=False):
+    def __init__(self, name, nt, shape, spc_border=0, forward=True, time_order=0, openmp=False, profile=False, cache_blocking=False, block_size=5):
         self.t = symbols("t")
+        self.cache_blocking = cache_blocking
+        try:
```
Comment (mlange05) [inline devito/propagator.py]: I don't think `try`/`except` is appropriate here. Why not check for list/tuples via `if isinstance(Iterable):`, with `from collections import Iterable`?

File: tests/test_cache_blocking.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        input_grid.data[:] = np.arange(90000, dtype=np.float64).reshape((300, 300))
+        output_grid_noblock = DenseData("output_grid", (300, 300), np.float64)
+        x, t = symbols("x t")
+        eq = Eq(output_grid_noblock[t, x], input_grid[t, x] + 3)
+        op_noblock = SimpleOperator(input_grid, output_grid_noblock, [eq])
+        op_noblock.apply()
+        output_grid_block = DenseData("output_grid", (300, 300), np.float64)
+        op_block = SimpleOperator(input_grid, output_grid_block, [eq], cache_blocking=True)
+        op_block.apply()
+        assert(np.linalg.norm(output_grid_block.data) == np.linalg.norm(output_grid_noblock.data))
```
Comment (mlange05) [inline tests/test_cache_blocking.py]: Why not compare the data arrays element-wise here? They should be identical, no? Same goes for the other test.

**Comment by mlange05** [comment]
Looks great. I suspect the `SimpleOperator` will go once we clean up the `Operator` interface in general, and we might be able to simplify the loop blocking itself once we introduce a more general `Loop` abstraction, but for now this looks very neat. Good to go once the last two comments are addressed.

File: tests/test_cache_blocking.py (hunk 3 above)
Comment (navjotk) [inline tests/test_cache_blocking.py]: Running assert inside a loop would create thousands of tests, right? Is there a numpy shortcut for element-wise comparison that I am missing?

Comment (mlange05) [inline tests/test_cache_blocking.py]: I believe `numpy.equal` should do the trick.

**Comment by mlange05** [comment]
Thanks, this looks great now. So, merging...

**Pull request merged**